main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -371,6 +373,195 @@ object Inliner {
       Expr.NewObject(name, clazz, tpe, eff, methods, loc)
   }
 
+  /**
+    * Evaluate match if possible.
+    *
+    * The match rules are iterated top to bottom, at each step either:
+    *   - Removing the rule if it can never match (either the patten cannot match or the guard is guaranteed false)
+    *   - Removing the match if a rule is known to match (either `Some(12)` matching `Some(x)` or a lenient pattern like `_`)
+    *   - Returning the rules as is if the rule cannot be determined statically.
+    */
+  @tailrec
+  private def reduceMatch(exp: MonoAst.Expr, rules: List[MonoAst.MatchRule], tpe: Type, eff: Type, loc: SourceLocation)(implicit sym0: Symbol.DefnSym, sctx: SharedContext): Expr = {
+    rules match {
+      case MonoAst.MatchRule(pat, guardOpt, ruleExp) :: rest =>
+        matchRule(exp, pat, guardOpt) match {
+          case MatchResult.Match(binders) =>
+            // Guaranteed rule - convert to let binders.
+            sctx.changed.putIfAbsent(sym0, ())
+            bindPatterns(binders.toSeq, ruleExp, loc)
+          case MatchResult.NoMatch =>
+            // Impossible rule - delete and continue.
+            sctx.changed.putIfAbsent(sym0, ())
+            reduceMatch(exp, rest, tpe, eff, loc)
+          case MatchResult.Unknown =>
+            // Match is unknown - do nothing.
+            Expr.Match(exp, rules, tpe, eff, loc)
+        }
+      case Nil =>
+        // Do nothing.
+        Expr.Match(exp, rules, tpe, eff, loc)
+    }
+  }
+
+  private sealed trait FuzzyBool
+
+  private object FuzzyBool {
+
+    case object True extends FuzzyBool
+
+    case object False extends FuzzyBool
+
+    case object Unknown extends FuzzyBool
+
+  }
+
+  /** Returns the fuzzy evaluation of `exp` as a boolean. */
+  private def evalBoolExpression(exp: MonoAst.Expr): FuzzyBool = exp match {
```

Comment:
I suppose its OK, but Option[Boolean] would also work right? Is it worth it. I don't see that you have any abstract operations on fuzzy bool like and or . So maybe not?

Reply:
Makes the data more self-documenting instead of relying so much on comments